docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,113 @@
+- Allow changes to device profile or device resource fields when the field is inconsequential to the object's use (ex: description) or when it is not yet associated to another EdgeX object.  These changes are listed as `Anytime` or `Situational` accordingly in the table below.
+- Allow the addition of new profiles, device resources, device commands, etc. **always**.  Additions may be accomplished via POST or PATCH depending on the change and API.
+
+See [Proposed Design section](#proposed-design) section below for specific metadata API changes.
+
+While not explicitly covered in this ADR, a device is not allowed to be deleted when it is associated to an event or reading.
+
+### Profile Changes
+
+The table below outlines the current elements of the EdgeX device profile, and the circumstances allowed for change (accomplished via PUT update API call - whether that is add, update or delete elements where applicable).
+
+| Device Profile Property | Type | Allowed Change | Description/Comments |
+|-------------------------|------|----------------|----------------------|
+| *General Properties* ||||
+| Id | string | Never | This uuid is generated by EdgeX and uniquely identifies the profile. |
+| Name | string | Never | This required property must be unique to the EdgeX instance and cannot be changed.  The profile name can never be empty and must be unique across the entire EdgeX instance. |
+| Manufacturer | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Description | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Model | string | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include setting the string to an empty string. |
+| Labels | string array | Anytime | As this property does not affect any functionality other than profile query, this property can be changed at any time – to include adding, removing or modifying any of the strings in the array. |
+| Device Resources | array | Situational | New device resources can always be added. Existing device resources can only be removed from the array when the device resource has not yet been associated to any reading or event.|
+| Device Commands | array | Situational | New device commands can always be added.  Existing device commands can only be removed from the array when the device command has not yet been associated to a reading or event.  Please see the device command properties for rules on modifying device commands.Device commands are optional and so the array can always be empty. |
```

Comment:
We should not search event and reading for modifying individual Device Resource or Device Command, because we should only allowed the changes when there is no Device associated to the Profile.
As the ADR mentioned, a device is not allowed to be deleted when it is associated to an event or reading, so checking is a more efficient way.
Checking event and reading is time-consuming, and the new event might be generated during the checking.

Reply:
It required the new API in core-data to just return the boolean for making sure the Event/Reading exists by source name and profile name.
Our current API will query and return lots of data.

Reply:
With the GET Reading/Device/Name{deviceName}/ResourceName{resourceName} API, you could request just 1 as a test, and see if you get 0 or 1. This is laborious, as has been noted, but seems like an existing way to answer the question.